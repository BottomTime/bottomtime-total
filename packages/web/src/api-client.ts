import { ApiClient } from '@bottomtime/api';

import { InjectionKey, inject } from 'vue';

export const ApiClientKey: InjectionKey<ApiClient> = Symbol('ApiClient');
export function useClient(): ApiClient {
  const client = inject(ApiClientKey);

  if (!client) {
    throw new Error(
      'Api Client has not been initialized. Did you forget to call `app.provide(key, client)`?',
    );
  }

  return client;
}
