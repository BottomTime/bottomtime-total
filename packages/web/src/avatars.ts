import { AvatarSize, ListAvatarURLsResponseDTO } from '@bottomtime/api';

export type AvatarURLs = ListAvatarURLsResponseDTO['sizes'];

export function getAvatarURL(
  baseUrl: string | undefined,
  size: AvatarSize,
): string | undefined {
  if (!baseUrl) return undefined;

  if (/^\/api\/users\/[a-z0-9]+([_.-][a-z0-9]+)*\/avatar/i.test(baseUrl)) {
    return baseUrl.endsWith('/') ? `${baseUrl}${size}` : `${baseUrl}/${size}`;
  }

  return baseUrl;
}
