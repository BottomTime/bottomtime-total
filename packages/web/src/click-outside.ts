import { Directive, DirectiveBinding } from 'vue';

type HTMLElementWithClickOutside = HTMLElement & {
  clickOutsideEvent: (e: MouseEvent) => void;
};

export const clickOutside: Directive = {
  beforeMount: (el: HTMLElementWithClickOutside, binding: DirectiveBinding) => {
    el.clickOutsideEvent = (event) => {
      // here I check that click was outside the el and his children
      if (!(el == event.target || el.contains(event.target as Node))) {
        // and if it did, call method provided in attribute value
        binding.value();
      }
    };
    document.addEventListener('click', el.clickOutsideEvent);
  },
  unmounted: (el: HTMLElementWithClickOutside) => {
    document.removeEventListener('click', el.clickOutsideEvent);
  },
};
