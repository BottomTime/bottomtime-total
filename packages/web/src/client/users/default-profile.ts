import { assertValid } from '@/helpers';
import { SuperAgentStatic } from 'superagent';
import {
  Profile,
  ProfileCertificationData,
  ProfileData,
  ProfileDataSchema,
  UserData,
} from './interfaces';
import { ProfileVisibility } from '@/constants';

export class DefaultProfile implements Profile {
  private static readonly JsonOmitKeys = new Set([
    'userId',
    'memberSince',
    'username',
  ]);

  constructor(
    private readonly agent: SuperAgentStatic,
    private readonly data: UserData,
  ) {}

  get userId(): string {
    return this.data.id;
  }
  get memberSince(): Date {
    return this.data.memberSince;
  }
  get username(): string {
    return this.data.username;
  }

  get avatar(): string | undefined {
    return this.data.profile.avatar;
  }
  set avatar(value: string | undefined) {
    this.data.profile.avatar = value;
  }

  get bio(): string | undefined {
    return this.data.profile.bio;
  }
  set bio(value: string | undefined) {
    this.data.profile.bio = value;
  }

  get birthdate(): string | undefined {
    return this.data.profile.birthdate;
  }
  set birthdate(value: string | undefined) {
    this.data.profile.birthdate = value;
  }

  get customData(): Record<string, unknown> | undefined {
    return this.data.profile.customData;
  }
  set customData(value: Record<string, unknown> | undefined) {
    this.data.profile.customData = value;
  }

  get certifications(): ProfileCertificationData[] | undefined {
    return this.data.profile.certifications;
  }
  set certifications(value: ProfileCertificationData[] | undefined) {
    this.data.profile.certifications = value;
  }

  get experienceLevel(): string | undefined {
    return this.data.profile.experienceLevel;
  }
  set experienceLevel(value: string | undefined) {
    this.data.profile.experienceLevel = value;
  }

  get location(): string | undefined {
    return this.data.profile.location;
  }
  set location(value: string | undefined) {
    this.data.profile.location = value;
  }

  get name(): string | undefined {
    return this.data.profile.name;
  }
  set name(value: string | undefined) {
    this.data.profile.name = value;
  }

  get profileVisibility(): ProfileVisibility {
    return this.data.profile.profileVisibility;
  }
  set profileVisibility(value: ProfileVisibility) {
    this.data.profile.profileVisibility = value;
  }

  get startedDiving(): string | undefined {
    return this.data.profile.startedDiving;
  }
  set startedDiving(value: string | undefined) {
    this.data.profile.startedDiving = value;
  }

  async save(): Promise<void> {
    const profile = assertValid<ProfileData>(
      this.data.profile,
      ProfileDataSchema,
    );
    await this.agent.put(`/api/profiles/${this.username}`).send(profile);
    this.data.profile = profile;
  }

  toJSON(): ProfileData {
    return assertValid(this.data.profile, ProfileDataSchema);
  }
}
