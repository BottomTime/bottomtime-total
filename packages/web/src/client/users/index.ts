export * from './interfaces';

export * from './default-user';
export * from './default-user-manager';
