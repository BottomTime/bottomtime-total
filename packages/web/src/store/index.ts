export * from './alerts.store';
export * from './current-user.store';
export * from './operators.store';
export * from './dive-sites.store';
export * from './error.store';
export * from './friends.store';
export * from './log-entries.store';
export * from './profiles.store';
export * from './tanks.store';
export * from './toasts.store';
