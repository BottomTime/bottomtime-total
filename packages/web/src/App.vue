<template>
  <div v-if="isLoading"></div>
  <div v-else>
    <SnackBar />
    <NavBar />
    <RouterView />
    <PageFooter />
    <CookieWarning />
  </div>
</template>

<script lang="ts" setup>
import { onBeforeMount, ref } from 'vue';

import CookieWarning from './components/main/CookieWarning.vue';
import { Dispatch, useStore } from '@/store';
import { inject } from '@/helpers';
import NavBar from '@/components/main/NavBar.vue';
import PageFooter from '@/components/main/PageFooter.vue';
import SnackBar from '@/components/main/SnackBar.vue';
import { ApiClientKey } from '@/injection-keys';

const client = inject(ApiClientKey);
const store = useStore();
const isLoading = ref(true);

onBeforeMount(async () => {
  try {
    const currentUser = await client.users.getCurrentUser();
    await store.dispatch(Dispatch.SignInUser, currentUser);
  } catch (error) {
    console.error(error);
  } finally {
    isLoading.value = false;
  }
});
</script>

<style>
@import './assets/bulmaswatch.min.css';
</style>
