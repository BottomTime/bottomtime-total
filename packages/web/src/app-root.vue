<template>
  <SnackBar />
  <div
    class="md:container mx-auto font-content text-grey-900 bg-blue-200 dark:text-grey-100 dark:bg-blue-950 rounded-b-xl shadow-md shadow-white dark:shadow-grey-700 opacity-95"
  >
    <section class="mt-16 p-4">
      <RouterView></RouterView>
    </section>
    <PageFooter />
  </div>
  <NavBar />
</template>

<script setup lang="ts">
import { onBeforeMount, onServerPrefetch, useSSRContext } from 'vue';

import { ToastType } from './common';
import NavBar from './components/core/nav-bar.vue';
import PageFooter from './components/core/page-footer.vue';
import SnackBar from './components/core/snack-bar.vue';
import { Config } from './config';
import { AppInitialState, useInitialState } from './initial-state';
import { useCurrentUser, useToasts } from './store';

const currentUser = useCurrentUser();
const ctx = Config.isSSR ? useSSRContext<AppInitialState>() : undefined;
const initialState = useInitialState();
const toasts = useToasts();

onServerPrefetch(async () => {
  currentUser.user = ctx?.currentUser ?? null;
});

onBeforeMount(() => {
  if (!Config.isSSR) {
    currentUser.user = initialState?.currentUser ?? null;

    if (initialState?.error) {
      if (!Config.isProduction) {
        /* eslint-disable-next-line no-console */
        console.error(initialState.error);
      }
      toasts.toast({
        id: 'server-prerender-error',
        message:
          'A server error has occurred and the page may not have loaded correctly. We apologize for the inconvenience. Please try again later as we investigate the issue.',
        type: ToastType.Error,
      });
    }
  }
});
</script>
