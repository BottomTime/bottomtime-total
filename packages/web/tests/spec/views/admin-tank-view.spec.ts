import {
  ApiClient,
  Fetcher,
  Tank,
  TankDTO,
  TankMaterial,
} from '@bottomtime/api';

import {
  ComponentMountingOptions,
  flushPromises,
  mount,
  renderToString,
} from '@vue/test-utils';

import { Pinia, createPinia } from 'pinia';
import { Router } from 'vue-router';

import { ApiClientKey } from '../../../src/api-client';
import { LocationKey, MockLocation } from '../../../src/location';
import { useCurrentUser, useTanks, useToasts } from '../../../src/store';
import AdminTankView from '../../../src/views/admin-tank-view.vue';
import { createHttpError } from '../../fixtures/create-http-error';
import { createRouter } from '../../fixtures/create-router';
import { AdminUser, BasicUser } from '../../fixtures/users';

const TestData: TankDTO = {
  id: '776c6a29-faf0-4577-9092-90fcf8e5c9c4',
  name: 'Test Tank',
  volume: 12,
  workingPressure: 232,
  material: TankMaterial.Steel,
  isSystem: true,
};

describe('Admin Tank View', () => {
  let fetcher: Fetcher;
  let client: ApiClient;
  let router: Router;

  let pinia: Pinia;
  let location: MockLocation;
  let currentUser: ReturnType<typeof useCurrentUser>;
  let tanks: ReturnType<typeof useTanks>;
  let toasts: ReturnType<typeof useToasts>;
  let opts: ComponentMountingOptions<typeof AdminTankView>;

  beforeAll(() => {
    fetcher = new Fetcher();
    client = new ApiClient({ fetcher });
    router = createRouter([
      {
        path: '/admin/tanks/:tankId',
        component: AdminTankView,
      },
    ]);
  });

  beforeEach(async () => {
    pinia = createPinia();
    currentUser = useCurrentUser(pinia);
    tanks = useTanks(pinia);
    toasts = useToasts(pinia);

    currentUser.user = AdminUser;
    tanks.currentTank = TestData;

    location = new MockLocation();

    await router.push(`/admin/tanks/${TestData.id}`);

    opts = {
      global: {
        plugins: [pinia, router],
        provide: {
          [ApiClientKey as symbol]: client,
          [LocationKey as symbol]: location,
        },
      },
    };
  });

  describe('when prefetching on the server side', () => {
    it('will retrieve the tank profile and render the form correctly', async () => {
      const spy = jest
        .spyOn(client.tanks, 'getTank')
        .mockResolvedValue(new Tank(fetcher, TestData));

      const raw = await renderToString(AdminTankView, { global: opts.global });
      const html = document.createElement('div');
      html.innerHTML = raw;

      expect(spy).toHaveBeenCalledWith(TestData.id);
      expect(html.querySelector<HTMLInputElement>('#name')?.value).toBe(
        TestData.name,
      );
      expect(html.querySelector<HTMLInputElement>('#volume')?.value).toBe(
        TestData.volume.toString(),
      );
      expect(html.querySelector<HTMLInputElement>('#pressure')?.value).toBe(
        TestData.workingPressure.toString(),
      );
      expect(
        html.querySelector<HTMLInputElement>('#material-fe')?.checked,
      ).toBe(true);
    });

    it('will show a not found message if the tank does not exist', async () => {
      const spy = jest
        .spyOn(client.tanks, 'getTank')
        .mockRejectedValue(createHttpError(404));

      const raw = await renderToString(AdminTankView, { global: opts.global });
      const html = document.createElement('div');
      html.innerHTML = raw;

      expect(spy).toHaveBeenCalledWith(TestData.id);
      expect(
        html.querySelector<HTMLParagraphElement>(
          '[data-testid="not-found-message"]',
        ),
      ).toBeDefined();
      expect(html.querySelector('#name')).toBeNull();
    });

    it('will show login form if user is not authenticated', async () => {
      const spy = jest
        .spyOn(client.tanks, 'getTank')
        .mockRejectedValue(createHttpError(401));
      currentUser.user = null;

      const raw = await renderToString(AdminTankView, { global: opts.global });
      const html = document.createElement('div');
      html.innerHTML = raw;

      expect(spy).toHaveBeenCalled();
      expect(html.querySelector('[data-testid="login-form"]')).toBeDefined();
      expect(html.querySelector('#name')).toBeNull();
    });

    it('will show forbidden message if user is not an admin', async () => {
      const spy = jest
        .spyOn(client.tanks, 'getTank')
        .mockRejectedValue(createHttpError(403));
      currentUser.user = BasicUser;

      const raw = await renderToString(AdminTankView, { global: opts.global });
      const html = document.createElement('div');
      html.innerHTML = raw;

      expect(spy).toHaveBeenCalled();
      expect(
        html.querySelector('[data-testid="forbidden-message"]'),
      ).toBeDefined();
      expect(html.querySelector('#name')).toBeNull();
    });
  });

  describe('when rendering on the client side', () => {
    it('will allow a user to update a tank profile', async () => {
      const expected: TankDTO = {
        ...TestData,
        name: 'New Name',
        volume: 15,
        workingPressure: 300,
        material: TankMaterial.Aluminum,
      };

      const tank = new Tank(fetcher, expected);
      jest.spyOn(client.tanks, 'wrapDTO').mockImplementation((dto) => {
        expect(dto).toEqual(expected);
        return tank;
      });
      const spy = jest.spyOn(tank, 'save').mockResolvedValue();

      const wrapper = mount(AdminTankView, opts);

      await wrapper.get('#name').setValue(expected.name);
      await wrapper.get('#volume').setValue(expected.volume.toString());
      await wrapper
        .get('#pressure')
        .setValue(expected.workingPressure.toString());
      await wrapper.get('#material-al').setValue(true);
      await wrapper.get('#save-tank').trigger('click');
      await flushPromises();

      expect(spy).toHaveBeenCalled();
      expect(toasts.toasts[0].message).toMatchSnapshot();
    });

    it('will prevent users form saving tanks if there are validation errors', async () => {
      const spy = jest.spyOn(client.tanks, 'wrapDTO');

      const wrapper = mount(AdminTankView, opts);

      await wrapper.get('#volume').setValue('nope');
      await wrapper.get('#save-tank').trigger('click');
      await flushPromises();

      expect(spy).not.toHaveBeenCalled();
      expect(toasts.toasts).toHaveLength(0);
      expect(
        wrapper.find('[data-testid="volume-error"]').text(),
      ).toMatchSnapshot();
    });

    it('will show a not found message if the tank does not exist', async () => {
      tanks.currentTank = null;
      const wrapper = mount(AdminTankView, opts);
      expect(
        wrapper.find('[data-testid="not-found-message"]').isVisible(),
      ).toBe(true);
      expect(wrapper.find('#name').exists()).toBe(false);
    });

    it('will show login form if user is not authenticated', async () => {
      currentUser.user = null;
      const wrapper = mount(AdminTankView, opts);
      expect(wrapper.find('[data-testid="login-form"]').isVisible()).toBe(true);
      expect(wrapper.find('#name').exists()).toBe(false);
    });

    it('will show forbidden message if user is not an admin', async () => {
      currentUser.user = BasicUser;
      const wrapper = mount(AdminTankView, opts);
      expect(
        wrapper.find('[data-testid="forbidden-message"]').isVisible(),
      ).toBe(true);
      expect(wrapper.find('#name').exists()).toBe(false);
    });
  });
});
