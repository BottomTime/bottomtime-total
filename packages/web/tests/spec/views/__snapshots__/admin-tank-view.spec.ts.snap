// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Admin Tank View when rendering on the client side will allow a user to update a tank profile 1`] = `"Tank profile saved successfully"`;

exports[`Admin Tank View when rendering on the client side will prevent users form saving tanks if there are validation errors 1`] = `"Volume must be between 1 and 30 litres"`;
